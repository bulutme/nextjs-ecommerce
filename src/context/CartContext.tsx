import React, {
  createContext,
  useContext,
  ReactNode,
  useEffect,
  useState,
} from "react";
import { ProductItemProps } from "@/app/types/Product/types";
import useLocalStorage from "@/hooks/useLocalStorage";

type CartItem = {
  product: ProductItemProps;
  count: number;
};

type GroupedCart = {
  brand: string;
  products: CartItem[];
};

type CartContextType = {
  cart: GroupedCart[];
  addToCart: (product: ProductItemProps) => void;
  removeProduct: (productId: number) => void;
  increaseCount: (productId: number) => void;
  decreaseCount: (productId: number) => void;
  totalItemCount: number;
  totalPrice: string;
  isLoading: boolean;
};

const CartContext = createContext<CartContextType | undefined>(undefined);

const CartProvider = ({ children }: { children: ReactNode }) => {
  const [cart, setCart] = useLocalStorage<GroupedCart[]>("cart", []);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    if (typeof window !== "undefined") {
      setIsLoading(true);
      setCart(JSON.parse(localStorage.getItem("cart") as string));
      setIsLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const totalItemCount = cart.reduce(
    (total, group) =>
      total +
      group.products.reduce((groupTotal, item) => groupTotal + item.count, 0),
    0
  );

  const totalPrice = cart.reduce(
    (total, group) =>
      total +
      group.products.reduce(
        (groupTotal, item) => groupTotal + item.count * item.product.price,
        0
      ),
    0
  );

  const formattedTotalPrice = totalPrice.toFixed(2);

  const addToCart = (product: ProductItemProps) => {
    const existingGroupIndex = cart.findIndex(
      (group) => group.brand === product.brand
    );

    if (existingGroupIndex !== -1) {
      const existingProductIndex = cart[existingGroupIndex].products.findIndex(
        (item) => item.product.id === product.id
      );

      const newCart = [...cart];

      if (existingProductIndex !== -1) {
        const cartId =
          newCart[existingGroupIndex].products[existingProductIndex].product.id;
        increaseCount(cartId);
      } else {
        newCart[existingGroupIndex].products.push({
          product,
          count: 1,
        });
        setCart(newCart);
      }
    } else {
      setCart((prevCart) => [
        ...prevCart,
        {
          brand: product.brand,
          products: [
            {
              product,
              count: 1,
            },
          ],
        },
      ]);
    }
  };

  const removeProduct = (productId: number) => {
    const newCart = cart.map((group) => ({
      ...group,
      products: group.products.filter((item) => item.product.id !== productId),
    }));

    // Remove empty groups
    const updatedCart = newCart.filter((group) => group.products.length > 0);
    setCart(updatedCart);
  };

  const increaseCount = (productId: number) => {
    const newCart = cart.map((group) => {
      return {
        ...group,
        products: group.products.map((item) =>
          item.product.id === productId
            ? { ...item, count: item.count + 1 }
            : item
        ),
      };
    });

    setCart(newCart);
  };

  const decreaseCount = (productId: number) => {
    const newCart = cart.map((group) => ({
      ...group,
      products: group.products.map((item) =>
        item.product.id === productId
          ? { ...item, count: Math.max(1, item.count - 1) }
          : item
      ),
    }));

    setCart(newCart);
  };

  const contextValue: CartContextType = {
    cart,
    totalItemCount,
    totalPrice: formattedTotalPrice,
    addToCart,
    removeProduct,
    increaseCount,
    decreaseCount,
    isLoading,
  };

  return (
    <CartContext.Provider value={contextValue}>{children}</CartContext.Provider>
  );
};

const useCart = () => {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error("useCart must be used within a CartProvider");
  }
  return context;
};

export { CartProvider, useCart };
