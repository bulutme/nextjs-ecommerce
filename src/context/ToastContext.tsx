import { createContext, FC, ReactNode, useContext, useState } from "react";

export type ToastType = "success" | "error" | "exit";

type ToastData = {
  id: number;
  type: ToastType;
  message: string;
};

type ToastContextType = {
  toasts: ToastData[];
  addToast: (type: ToastType, message: string) => void;
  removeToast: (id: number) => void;
};

const MAX_TOASTS = 5;

const ToastContext = createContext<ToastContextType | undefined>(undefined);

export const ToastProvider: FC<{ children: ReactNode }> = ({ children }) => {
  const [toasts, setToasts] = useState<ToastData[]>([]);

  // function to add toast using type and message
  const addToast = (type: ToastType, message: string) => {
    const id = new Date().getTime();
    const newToast: ToastData = { id, type, message };

    // remove the latest one if toast item more than 5
    setToasts((prevToasts) => [...prevToasts.slice(-MAX_TOASTS + 1), newToast]);

    if (type !== "exit") {
      setTimeout(() => {
        removeToast(id);
      }, 3000);
    }
  };

  // function to remove toast
  const removeToast = (id: number) => {
    setToasts((prevToasts) => prevToasts.filter((toast) => toast.id !== id));
  };

  const contextValue: ToastContextType = {
    toasts,
    addToast,
    removeToast,
  };

  return (
    <ToastContext.Provider value={contextValue}>
      {children}
    </ToastContext.Provider>
  );
};

// custom hook to use toast context
export const useToast = () => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error("useToast must be used within a ToastProvider");
  }
  return context;
};
