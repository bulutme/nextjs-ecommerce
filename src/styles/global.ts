"use client";

import { createGlobalStyle } from "styled-components";

const GlobalStyles = createGlobalStyle`
  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }

  html {
    font-size: 16px;
  }

  html, body, #__next {
    height: 100%;
  }

  a {
    text-decoration: none;
    width: fit-content;
  }

  ul {
    list-style-type: none
  }
`;

export default GlobalStyles;
