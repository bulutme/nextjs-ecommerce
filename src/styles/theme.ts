export const theme = {
  colors: {
    black: "#000",
    white: "#fff",
    gray: "#808080",
    grayLight: "#d3d3d3",
    border: "#e2e8f0",
    primary: "#3bb77e",
    primaryDark: "#2b8a5e",
    error: "#dc2626",
    info: "#0891b2",
    success: "#16a34a",
  },
  screens: {
    "2xs": "400px",
    xs: "480px",
    sm: "576px",
    md: "768px",
    lg: "992px",
    xl: "1200px",
    "2xl": "1440px",
    "3xl": "1560px",
  },
  fontWeight: {
    normal: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
  },
};
