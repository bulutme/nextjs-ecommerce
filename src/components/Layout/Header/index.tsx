import { FC, useCallback } from "react";
import styled from "styled-components";
import Image from "next/image";
import Button from "@/components/Button";
import SearchInput from "@/components/Search";
import { MdOutlineShoppingCart } from "react-icons/md";
import Link from "next/link";
import Container from "@/components/Container";
import CartPopup from "@/app/cart/components/Cart/CartPopup";
import { usePathname, useRouter, useSearchParams } from "next/navigation";
import { useCart } from "@/context/CartContext";
import { AiOutlineArrowLeft } from "react-icons/ai";

const HeaderContainer = styled.div`
  position: relative;
  width: 100%;
  padding-top: 24px;
  padding-bottom: 8px;
  border-bottom: 1px solid ${({ theme }) => theme.colors.border};
  max-height: auto;
  padding: 0.6rem;
  display: flex;
  flex-direction: column;
  gap: 1rem;

  @media (min-width: ${({ theme }) => theme.screens.xs}) {
    padding: 1rem;
  }

  @media (min-width: ${({ theme }) => theme.screens.md}) {
    justify-content: space-around;
    gap: 7%;
    align-items: center;
    flex-direction: row;
    margin-bottom: 2rem;
  }
`;

const ButtonWrapper = styled.div`
  position: relative;
`;

const Cart = styled.div`
  position: absolute;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  width: 200px;
  right: 32px;
  top: 18px;
  height: 40px;

  @media (min-width: ${({ theme }) => theme.screens.md}) {
    position: static;
  }
`;

const Badge = styled.span<{ $showBadge: boolean }>`
  position: absolute;
  top: 0;
  right: -10px;
  background-color: ${({ theme }) => theme.colors.error};
  text-align: center;
  color: ${({ theme }) => theme.colors.white};
  font-size: 0.75rem;
  font-weight: ${({ theme }) => theme.fontWeight.semibold};
  line-height: 22px;
  border-radius: 50%;
  width: 22px;
  height: 22px;
  display: ${({ $showBadge }) => ($showBadge ? "block" : "none")};
`;

const Header: FC = () => {
  const router = useRouter();
  const pathname = usePathname();
  const searchParams = useSearchParams();
  const { totalItemCount } = useCart();

  const createQueryString = useCallback(
    (name: string, value: string) => {
      const params = new URLSearchParams(searchParams);
      params.set(name, value);

      return params.toString();
    },
    [searchParams]
  );

  const handleChangeSearch = (value: string) => {
    router.push(
      value ? `/?${createQueryString("query", value)}` : pathname + ""
    );
  };

  const isCartVisible = pathname !== "/cart";

  return (
    <Container>
      <HeaderContainer>
        <Link href="/">
          <Image
            src={"/next.svg"}
            width={100}
            height={55}
            alt="next logo"
            priority
          />
        </Link>
        <SearchInput
          initialValue={searchParams.get("query")!}
          $fullwidth
          onSearch={(value) => handleChangeSearch(value)}
        />
        <Cart>
          {isCartVisible ? (
            <CartPopup>
              <Link href="/cart">
                <ButtonWrapper>
                  <Button
                    icon={<MdOutlineShoppingCart />}
                    type="button"
                    size="small"
                    $variant="link"
                    iconSize={24}
                  >
                    My Cart
                  </Button>
                  <Badge $showBadge={totalItemCount > 0}>
                    {totalItemCount}
                  </Badge>
                </ButtonWrapper>
              </Link>
            </CartPopup>
          ) : (
            <ButtonWrapper>
              <Link href="/">
                <Button
                  icon={<AiOutlineArrowLeft />}
                  size="small"
                  $variant="link"
                >
                  Proceed Shopping
                </Button>
              </Link>
            </ButtonWrapper>
          )}
        </Cart>
      </HeaderContainer>
    </Container>
  );
};

export default Header;
