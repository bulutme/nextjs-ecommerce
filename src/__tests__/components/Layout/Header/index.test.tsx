/* eslint-disable react/display-name */
/* eslint-disable @next/next/no-img-element */
import { render, screen } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { theme } from "@/styles/theme";
import Header from "@/components/Layout/Header";
import { CartProvider } from "@/context/CartContext";

jest.mock(
  "next/link",
  () =>
    ({ children }: { children: React.ReactNode }) =>
      children
);
jest.mock("next/image", () => ({ src, alt }: { src: string; alt: string }) => (
  <img src={src} alt={alt} />
));

const mockRouter = {
  push: jest.fn(),
  back: jest.fn(),
};
const mockUseSearchParams = new URLSearchParams();
jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(() => mockUseSearchParams),
  usePathname: jest.fn(),
  useRouter: () => mockRouter,
}));

const mockUseCart = {
  totalItemCount: 2,
};
jest.mock("../../../../context/CartContext", () => ({
  useCart: jest.fn(() => mockUseCart),
}));

describe("Header component", () => {
  test("renders Header component", () => {
    render(
      <ThemeProvider theme={theme}>
        <Header />
      </ThemeProvider>
    );

    expect(screen.getByAltText("next logo")).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText("Search by product, brand and category")
    ).toBeInTheDocument();
    expect(screen.getByText("My Cart")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Header />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  test("renders Header component with different data", () => {
    const mockUseCart = {
      totalItemCount: 5,
    };
    jest.mock("../../../../context/CartContext", () => ({
      useCart: jest.fn(() => mockUseCart),
    }));

    render(
      <ThemeProvider theme={theme}>
        <Header />
      </ThemeProvider>
    );

    expect(screen.getByAltText("next logo")).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText("Search by product, brand and category")
    ).toBeInTheDocument();
    expect(screen.getByText("My Cart")).toBeInTheDocument();
  });

  test("matches snapshot with different data", () => {
    const mockUseCart = {
      totalItemCount: 5,
    };
    jest.mock("../../../../context/CartContext", () => ({
      useCart: jest.fn(() => mockUseCart),
    }));

    const { container } = render(
      <ThemeProvider theme={theme}>
        <Header />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
