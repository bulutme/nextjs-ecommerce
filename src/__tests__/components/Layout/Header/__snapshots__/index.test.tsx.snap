// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component matches snapshot 1`] = `
<div>
  <div
    class="sc-f2eef38d-0 iwNFqL"
  >
    <div
      class="sc-bc8a6216-0 doMBAj"
    >
      <img
        alt="next logo"
        src="/next.svg"
      />
      <div
        class="sc-eb9f1157-0 dMzKcq"
      >
        <input
          class="sc-acef6806-0 bmACrq"
          name="search"
          placeholder="Search by product, brand and category"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-bc8a6216-2 jmsrXH"
      >
        <div
          class="sc-ec3ca1e3-0 gIMvgl"
        >
          <div
            class="sc-ec3ca1e3-1 hJnOck"
          >
            <div
              class="sc-bc8a6216-1 eAFvoD"
            >
              <button
                class="sc-e6c76d24-0 gigtzz"
                type="button"
              >
                <span
                  style="width: 24px; height: 24px; font-size: 24px; margin-right: 8px;"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0V0z"
                      fill="none"
                    />
                    <path
                      d="M15.55 13c.75 0 1.41-.41 1.75-1.03l3.58-6.49A.996.996 0 0020.01 4H5.21l-.94-2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2h7.45zM6.16 6h12.15l-2.76 5H8.53L6.16 6zM7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </span>
                My Cart
              </button>
              <span
                class="sc-bc8a6216-3 fXSUgL"
              >
                2
              </span>
            </div>
          </div>
          <div
            class="sc-ec3ca1e3-2 dNnrhF"
          >
            <div
              class="sc-871be3e9-0 jNPDlq"
            >
              <h4
                class="sc-871be3e9-1 lpgGan"
              >
                My Cart (
                2
                 Products)
              </h4>
              <div
                class="sc-871be3e9-2 bfXrOa"
              />
              <div
                class="sc-871be3e9-3 ldvweX"
              >
                <div
                  class="sc-871be3e9-4 fsMaVb"
                >
                  <span
                    class="sc-871be3e9-5 jHwUMS"
                  >
                    Total
                  </span>
                  <span
                    class="sc-871be3e9-6 dbNgkr"
                  >
                    $
                  </span>
                </div>
                <button
                  class="sc-e6c76d24-0 lfwATC"
                  color="primary"
                >
                  Go to Cart
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header component matches snapshot with different data 1`] = `
<div>
  <div
    class="sc-f2eef38d-0 iwNFqL"
  >
    <div
      class="sc-bc8a6216-0 doMBAj"
    >
      <img
        alt="next logo"
        src="/next.svg"
      />
      <div
        class="sc-eb9f1157-0 dMzKcq"
      >
        <input
          class="sc-acef6806-0 bmACrq"
          name="search"
          placeholder="Search by product, brand and category"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-bc8a6216-2 jmsrXH"
      >
        <div
          class="sc-ec3ca1e3-0 gIMvgl"
        >
          <div
            class="sc-ec3ca1e3-1 hJnOck"
          >
            <div
              class="sc-bc8a6216-1 eAFvoD"
            >
              <button
                class="sc-e6c76d24-0 gigtzz"
                type="button"
              >
                <span
                  style="width: 24px; height: 24px; font-size: 24px; margin-right: 8px;"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0V0z"
                      fill="none"
                    />
                    <path
                      d="M15.55 13c.75 0 1.41-.41 1.75-1.03l3.58-6.49A.996.996 0 0020.01 4H5.21l-.94-2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2h7.45zM6.16 6h12.15l-2.76 5H8.53L6.16 6zM7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </span>
                My Cart
              </button>
              <span
                class="sc-bc8a6216-3 fXSUgL"
              >
                2
              </span>
            </div>
          </div>
          <div
            class="sc-ec3ca1e3-2 dNnrhF"
          >
            <div
              class="sc-871be3e9-0 jNPDlq"
            >
              <h4
                class="sc-871be3e9-1 lpgGan"
              >
                My Cart (
                2
                 Products)
              </h4>
              <div
                class="sc-871be3e9-2 bfXrOa"
              />
              <div
                class="sc-871be3e9-3 ldvweX"
              >
                <div
                  class="sc-871be3e9-4 fsMaVb"
                >
                  <span
                    class="sc-871be3e9-5 jHwUMS"
                  >
                    Total
                  </span>
                  <span
                    class="sc-871be3e9-6 dbNgkr"
                  >
                    $
                  </span>
                </div>
                <button
                  class="sc-e6c76d24-0 lfwATC"
                  color="primary"
                >
                  Go to Cart
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
